### PR TITLE
Theme can now be changed by changing the theme prop

### PR DESCRIPTION
**Pull request by hetanthakkar1**
What kind of change does this PR introduce?

Added ability to change the theme for a component tree via state change
Fixed issue #2754

### CODE REVIEW

**Comment by hetanthakkar1** [comment]
@flyingcircle Sorry for reopening this PR again. I have made the necessary changes and made the code much simpler. Can you please review!? I closed that PR because I accidentally messed up with git and deleted that branch. Will be cautious next time. Sorry for trouble

File: src/config/ThemeProvider.tsx [hunk 1]
```diff
@@ -54,7 +54,10 @@ export default class ThemeProvider extends React.Component<
 
   static getDerivedStateFromProps(props, state) {
     const { useDark } = props;
-    if (useDark !== state.useDark) {
+    const isTheme = (theme) => {
+      return !(Object.keys(theme).length === 0 && theme.constructor === Object);
```
Comment (flyingcircle) [inline src/config/ThemeProvider.tsx:58]: I'm not sure I understand why this function? And I don't know what `isTheme` is even supposed to mean. What is this function trying to look for? That the theme is present? I thought we were supposed to look for whether the `theme` object changed or not.

**Review by flyingcircle — CHANGES_REQUESTED** [review]

**Comment by hetanthakkar1** [comment]
@flyingcircle  yeah! you are right. We need to detect that. I think I need to get some sleep. I am making too many silly mistakes.

**Comment by flyingcircle** [comment]
CI is saying you're failing the unit tests. Please fix first.

**Comment by hetanthakkar1** [comment]
@flyingcircle  I have also added `isTheme` condition in the if statement to check if the theme has been supplied. It is necessary to check as a test case in test file was failing as its style got changed in getDerivedStateFromProps because `state.theme!==props.theme` always holds true in the beginning if the theme is not supplied

**Comment by hetanthakkar1** [comment]
@flyingcircle I have also added `isTheme` condition in the if statement to check if the theme has been supplied. It is necessary to check as a test case in test file was failing as its style got changed in getDerivedStateFromProps because `state.theme!==props.theme` always holds true in the beginning if the theme is not supplied

File: src/config/ThemeProvider.tsx [hunk 2]
```diff
@@ -54,7 +54,13 @@ export default class ThemeProvider extends React.Component<
 
   static getDerivedStateFromProps(props, state) {
     const { useDark } = props;
-    if (useDark !== state.useDark) {
+    const isTheme = (theme) => {
```
Comment (flyingcircle) [inline src/config/ThemeProvider.tsx:57]: This function still doesn't make sense to me. Are we just checking `theme !== {}`. That seems a bit easier to understand than this.

**Review by flyingcircle — CHANGES_REQUESTED** [review]

File: src/config/ThemeProvider.tsx (hunk 2 above)
Comment (hetanthakkar1) [reply src/config/ThemeProvider.tsx:57]: Yeah! we are checking that if theme function is not empty

Comment (hetanthakkar1) [reply src/config/ThemeProvider.tsx:57]: That function is a more formal way of checking the same thing precisely

Comment (flyingcircle) [reply src/config/ThemeProvider.tsx:57]: Easy to read code is preferable.

Comment (hetanthakkar1) [reply src/config/ThemeProvider.tsx:57]: shall i replace that by theme !== {}?

**Comment by hetanthakkar1** [comment]
@flyingcircle  Please review!

File: src/config/ThemeProvider.tsx [hunk 3]
```diff
@@ -54,12 +54,10 @@ export default class ThemeProvider extends React.Component<
...
   static getDerivedStateFromProps(props, state) {
     const { useDark } = props;
-    const isTheme = (theme) => {
-      return !(Object.keys(theme).length === 0 && theme.constructor === Object);
-    };
+
     if (
       useDark !== state.useDark ||
-      (isTheme(props.theme) && props.theme !== state.theme)
+      (props.theme !== {} && props.theme !== state.theme)
```
Comment (arpitBhalla) [inline src/config/ThemeProvider.tsx:60]: We can't compare Objects in js

File: src/config/ThemeProvider.tsx (hunk 2 above)
Comment (flyingcircle) [reply src/config/ThemeProvider.tsx]: @hetanthakkar1 I see that I was wrong on this one. Can you change it back and just add a comment explaining the code?

**Comment by hetanthakkar1** [comment]
@flyingcircle Reverted the changes

**Comment by hetanthakkar1** [comment]
@flyingcircle Please review! I have made the necassary changes

File: src/config/ThemeProvider.tsx (hunk 2 above)
Comment (hetanthakkar1) [reply src/config/ThemeProvider.tsx:57]: @flyingcircle  done!

**Comment by hetanthakkar1** [comment]
@flyingcircle Can you please review!

**Comment by anthowm** [comment]
This will be available in next release ?

**Pull request merged**